[OpenMP][flang] Lower worksharing loops

SouraVX opened the pull request:
Still draft

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -55,8 +55,12 @@ static void createBodyOfOp(Op &op, Fortran::lower::FirOpBuilder &firOpBuilder,
   firOpBuilder.setInsertionPointToStart(&block);
-  // Ensure the block is well-formed.
-  firOpBuilder.create<mlir::omp::TerminatorOp>(loc);
+  if (op.getOperationName() == "omp.wsloop") {
```
Is there another way to do the comparison to avoid using a hard-coded string? Like `isa` or something like that.

schweitzpgi replied:
Yes, using `isa<FooOp>` is the preferred approach here.

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+    }
+  }
+  auto &doStmtEval = eval.getFirstNestedEvaluation();
+  auto *doStmt = doStmtEval.getIf<Fortran::parser::NonLabelDoStmt>();
```
You might want to check if this succeed.

SouraVX replied:
Thanks for review @clementval 
Yes, I've another dev version which have this early check. This is the first draft, some clauses and tests are pending.
Waiting on the next Main to FIR rebase.

clementval reviewed (commented):
I just had a quick look. Will come back on it later. Do you have any tests for this PR?

kiranchandramohan commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+    Fortran::lower::StatementContext stmtCtx;
+    lowerBound.push_back(fir::getBase(converter.genExprValue(
+        *Fortran::semantics::GetExpr(bounds->lower), stmtCtx)));
+    upperBound.push_back(fir::getBase(converter.genExprValue(
```
At the moment the lowering of the wsloop is not inclusive of the upperbound. We have to add an attribute (inclusive?) to the operation to include the upperbound.

kiranchandramohan replied:
The attribute is now in upstream llvm-project (see commit below). You have to either use that attribute or create an expression to increment the upperbound.
https://github.com/llvm/llvm-project/commit/268ff38a716157c362b8d463e2e5655f25972e42

kiranchandramohan replied:
@schweitzpgi are there plans for a merge from upstream soon? If not, is it possible to cherry-pick the following commit?https://github.com/llvm/llvm-project/commit/268ff38a716157c362b8d463e2e5655f25972e42

schweitzpgi replied:
The next rebase should be happening later this week.

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -125,30 +129,22 @@ genOMP(Fortran::lower::AbstractConverter &converter,
-    const auto &parallelOpClauseList =
-        std::get<Fortran::parser::OmpClauseList>(beginBlockDirective.t);
-    for (const auto &clause : parallelOpClauseList.v) {
+    const auto &OpClauseList =
```
Should be `opClauseList` to conform to the MLIR camelCase style here.

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+    auto &firOpBuilder = converter.getFirOpBuilder();
+    auto currentLocation = converter.getCurrentLocation();
     auto masterOp =
         firOpBuilder.create<mlir::omp::MasterOp>(currentLocation, argTy);
```
Is this `MasterOp` ever going to have arguments? (Not sure why this one needs an empty list of argument types.)

kiranchandramohan replied:
I don't think it is going to have arguments. Also, the Master construct is deprecated in OpenMP 5.1.
There is a build function without arguments. I don't know whether @SouraVX faced any issues with using it.

```
void MasterOp::build(::mlir::OpBuilder &odsBuilder, ::mlir::OperationState &odsState) {
  (void)odsState.addRegion();
}          
```

SouraVX replied:
I've noted down the comment when it was made, but unfortunately didn't paid much attention to it since this patch was my primary priority. I'll take a look at this again, a bit later.

kiranchandramohan replied:
OK with me to address in a separate patch.

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+      firstPrivateClauseOperands, lastPrivateClauseOperands, linearVars,
+      linearStepVars;
+  mlir::Value scheduleChunkClauseOperand;
+  Attribute scheduleClauseOperand, collapseClauseOperand, noWaitClauseOperand,
```
Are these `mlir::Attribute`? Might want to be explicit with a namespace prefix.

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+    createParallelOp<Fortran::parser::OmpBeginLoopDirective>(
+        converter, eval,
+        std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t));
+  else {
```
I prefer seeing both the `if` and the `else` be balanced with respect to both either having or eliding the `{ }` braces.

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -260,13 +254,101 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+      collapseClauseOperand.dyn_cast_or_null<IntegerAttr>(),
+      noWaitClauseOperand.dyn_cast_or_null<UnitAttr>(),
+      orderedClauseOperand.dyn_cast_or_null<IntegerAttr>(),
+      orderClauseOperand.dyn_cast_or_null<StringAttr>());
```
That's a long argument list. Has a single attribute dictionary argument been considered to slim this down a bit?

SouraVX replied:
That can done, however in `omp` lowering we follow this long list approach(for instance parallelOp also follows this). One benefit of this approach is that user/reader doesn't have to look anywhere else for arguments since `Op` is constructed there it self(using explicit arguments).

SouraVX commented:
Changes:
- Inclusive bounds not handled.
- Handle most clauses, specially `nowait` clause which appear at bottom i.e
```
!$OMP DO
<...>
!$OMP END DO NOWAIT
```
This warrants removal of 
https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Lower/Bridge.cpp#L1169 , Since clause part of end Loop Directive(such as `NOWAIT`) can be captured in LoopConstruct it self.
@ichoyjx, @schweitzpgi , @clementval @kiranchandramohan you have any thoughts WRT this ?

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -55,8 +55,12 @@ static void createBodyOfOp(Op &op, Fortran::lower::FirOpBuilder &firOpBuilder,
+  if (isa<omp::WsLoopOp>(op.getParentOp())) {
+    mlir::ValueRange results;
+    firOpBuilder.create<mlir::omp::YieldOp>(loc, results);
+  } else
```
You should add braces for the `else` block to keep it uniform with the `if` block. 

https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -266,13 +260,164 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+  const auto &blockDirective =
+      std::get<Fortran::parser::OmpBlockDirective>(beginBlockDirective.t);
+
+  if (blockDirective.v == llvm::omp::OMPD_parallel)
```
`if` block should have braces to be uniform with the `else` block. 

https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -266,13 +260,164 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+      noWaitClauseOperand, orderedClauseOperand, orderClauseOperand;
+  const auto &wsLoopOpClauseList = std::get<Fortran::parser::OmpClauseList>(
+      std::get<Fortran::parser::OmpBeginLoopDirective>(loopConstruct.t).t);
+  if (llvm::omp::OMPD_parallel_do ==
```
Same here for the braces

clementval commented:
> This warrants removal of
> https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Lower/Bridge.cpp#L1169 , Since clause part of end Loop Directive(such as `NOWAIT`) can be captured in LoopConstruct it self.
> @ichoyjx, @schweitzpgi , @clementval @kiranchandramohan you have any thoughts WRT this ?

I would handle the clauses on the `end` directive directly when you lower the begin directive as you did in your draft code.

SouraVX commented:
> > This warrants removal of
> > https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Lower/Bridge.cpp#L1169 , Since clause part of end Loop Directive(such as `NOWAIT`) can be captured in LoopConstruct it self.
> > @ichoyjx, @schweitzpgi , @clementval @kiranchandramohan you have any thoughts WRT this ?
> 
> I would handle the clauses on the `end` directive directly when you lower the begin directive as you did in your draft code.

Thanks @clementval. Let others also have a look, then I'll do a removal patch on Phabricator.

kiranchandramohan commented:
> TBD plan @kiranchandramohan:
> Add test cases when custom printing/parsing is in place(to avoid duplicate efforts). In meantime we can invest time in lowering.
> 

I think you can go ahead with a generic syntax trivial loop to demonstrate the lowering from pft to mlir and to llvm. We will update after the pretty printer is in.

> FYI: you can try `$bbc -fopenmp -emit-fir test.f90` and see the resultant FIR with generic `omp.wsloop` op. Obviously Lowering to LLVM cannot be done till #454 is fixed/handled.

https://github.com/flang-compiler/f18-llvm-project/pull/576 fixes the issue.
> 
> Changes:
> 
> * Inclusive bounds not yet handled.

This should come in the next merge or it should be cherry-picked.  https://github.com/llvm/llvm-project/commit/268ff38a716157c362b8d463e2e5655f25972e42

> * Handle most clauses, specially `nowait` clause which appear at bottom i.e
> 
> ```
> !$OMP DO
> <...>
> !$OMP END DO NOWAIT
> ```
> 
> This warrants removal of
> https://github.com/flang-compiler/f18-llvm-project/blob/fir-dev/flang/lib/Lower/Bridge.cpp#L1169 , Since clause part of end Loop Directive(such as `NOWAIT`) can be captured in LoopConstruct it self.
> @ichoyjx, @schweitzpgi , @clementval @kiranchandramohan you have any thoughts WRT this ?

Hope this is now fixed.

Will be great if we can merge this PR before the end of this week.

SouraVX commented:
Changes:
- Incorporated suggested changes.
- Added simple(useful) test case(If possible please try using `bbc` to get full taste of lowering).
Note: We have reached a decent patch size level. There may be room for  more refinement, however that can be done incrementally. What do you guys think ?

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -266,13 +261,175 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+          std::get<Fortran::parser::OmpScheduleClause::ScheduleType>(
+              scheduleClause->t);
+      switch (scheduleKind) {
+      default:
```
Do you cover all cases in this switch? If so, the default should not be present. 
https://llvm.org/docs/CodingStandards.html#don-t-use-default-labels-in-fully-covered-switches-over-enumerations

kiranchandramohan commented:
```
Added simple(useful) test case(If possible please try using bbc to get full taste of lowering).
Note: We have reached a decent patch size level. There may be room for more refinement, however that can be done incrementally. What do you guys think ?
```

OK with me.

kiranchandramohan reviewed (approved):
LGTM.

We should add the inclusive attribute as soon as it is available and add llvm ir generation tests.

SouraVX commented:
Changes done:
- Rebased (Prepared for merge).
- Adapted `inclusive` from new op definition.
- Adapted changes to `schedule` clause after @clementval upstream refactoring.
Miscellaneous:
- After recent rebase, `getParentOp` got removed (Can't find the commit). As a result operation classification is reverted to previous working iterarion.
i.e (Added FIXME for reminding)
```
-  if (isa<omp::WsLoopOp>(op.getParentOp())) {   // Doesn't work now
+  if (StringRef("omp.wsloop") == op.getOperationName()) {
```
Thank you all for your patience and valuable time!

SouraVX commented:
Rebased again.
Changes:
- String based checking removed i.e(StringRef("omp.wsloop") == op.getOperationName), Thanks to @kiranchandramohan!
- Test case further extended to capture full lowering to LLVMIR. (Executable can be generated)

kiranchandramohan commented:
@clementval Are you OK with this patch now?

clementval commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -267,13 +268,176 @@ genOMP(Fortran::lower::AbstractConverter &converter,
+      }
+    }
+  }
+  // In FORTRAN `nowait`clause occur at the end of `omp do` directive.
```
Missing space
```suggestion
  // In FORTRAN `nowait` clause occur at the end of `omp do` directive.
```

clementval commented:
> @clementval Are you OK with this patch now?

LGTM
